contracts/MultisigCaller.sol

```diff
@@ -0,0 +1,285 @@
+
+contract MultisigCaller is AccessControlEnumerable, ReentrancyGuard {
+    bytes32 public constant APPROVER_ROLE = keccak256("APPROVER_ROLE");
+    bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
+    uint256 public constant MIN_APPROVERS = 2;
+    uint256 public constant MAX_APPROVERS = 10;
+    
+    uint256 public requiredApprovals;
+
+    struct Transaction {
+        address to;
+        uint256 value;
+        bytes data;
+        bool executed;
+        uint256 approvalCount;
+    }
+
+    Transaction[] public transactions;
+    mapping(uint256 => mapping(address => bool)) public approvals;
+
+    event TransactionSubmitted(uint256 indexed txId, address indexed to, uint256 value, bytes data);
+    event TransactionApproved(uint256 indexed txId, address indexed approver);
+    event TransactionExecuted(uint256 indexed txId);
+    event ApproverAdded(address indexed account);
+    event ApproverRemoved(address indexed account);
+    event RequiredApprovalsChanged(uint256 oldRequired, uint256 newRequired);
+
+    using Address for address;
+
+    struct Call3 {
+        address target;
+        bool allowFailure;
+        bytes callData;
+    }
+
+    struct Call3Value {
+        address target;
+        bool allowFailure;
+        uint256 value;
+        bytes callData;
+    }
+
+    struct Result {
+        bool success;
+        bytes returnData;
+    }
+
+    constructor(address[] memory approvers, uint256 _requiredApprovals) {
+        if (approvers.length < MIN_APPROVERS) revert InvalidApproverCount(approvers.length, MIN_APPROVERS);
+        if (_requiredApprovals > approvers.length) revert RequiredApprovalsExceedApprovers(_requiredApprovals, approvers.length);
+        if (_requiredApprovals < MIN_APPROVERS) revert RequiredApprovalsTooLow(_requiredApprovals, MIN_APPROVERS);
+
+        requiredApprovals = _requiredApprovals;
+
+        // Setup the contract itself as the admin
+        _grantRole(DEFAULT_ADMIN_ROLE, address(this));
+        _grantRole(ADMIN_ROLE, address(this));
+
+        // Setup initial approvers
+        for (uint256 i = 0; i < approvers.length; i++) {
+            if (approvers[i] == address(0)) revert ZeroAddress();
+            _grantRole(APPROVER_ROLE, approvers[i]);
+        }
+    }
+
+    function submitTransaction(address to, uint256 value, bytes memory data) external payable nonReentrant onlyRole(APPROVER_ROLE) {
+        uint256 txId = transactions.length;
+        
+        Transaction memory newTx = Transaction({
+            to: to,
+            value: value,
+            data: data,
+            executed: false,
+            approvalCount: 0
+        });
+        transactions.push(newTx);
+        
+        emit TransactionSubmitted(txId, to, value, data);
+
+        Transaction storage transaction = transactions[txId];
+        _approveTransaction(transaction, txId);
+    }
+
+    function approveTransaction(uint256 txId) public payable nonReentrant onlyRole(APPROVER_ROLE) {
```

Comment:
should be external

Reply:
Done in 5c8331ad3f4ec10d3d2183419af2965fbd38f3f4.